ocaml/xapi-idl/lib/posix_channel.ml

```diff
@@ -80,12 +80,17 @@ let proxy (a : Unix.file_descr) (b : Unix.file_descr) =
       in
       (* If we can't make any progress (because fds have been closed), then stop *)
       if r = [] && w = [] then raise End_of_file ;
-      let r, w, _ = Unix.select r w [] (-1.0) in
-      (* Do the writing before the reading *)
-      List.iter
-        (fun fd -> if a = fd then CBuf.write b' a else CBuf.write a' b)
-        w ;
-      List.iter (fun fd -> if a = fd then CBuf.read a' a else CBuf.read b' b) r ;
+      let epoll = Polly.create () in
+      List.iter (fun fd -> Polly.add epoll fd Polly.Events.inp) (r @ w) ;
+      ignore
+      @@ Polly.wait epoll 4 (-1) (fun _ fd _ ->
```

Comment:
I think that the old code can be summarised as: "Wait for any `r` to be ready to reading or any of the `w` to be ready for writing. Then read/write to the fd found ready. Start again.". In every round of the loop (every time `select` returns) one or more fds would be ready, and it may often be just one. So we can't assume that we can write + read in the same round.

The third argument of the function passed in here, currently ignore with `_`, is set to the kind of event that happened and indicates whether we can read from or write to `fd`.